Arch arm cortex m direct dynamic irqs

ioannisg opened the pull request:
DIrect Dynamic Interrupt Feature support for ARM Cortex-M

ioannisg committed:
tests: arch: arm: move ZLI test to arm_irq_advanced_features suite

Move the zero-latency IRQ test into the new
arm_irq_advanced_features' test suite. Skip
running the test for non Mainline Cortex-M.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
tests: arch: arm: add test-suite for dynamic direct IRQs

We add a test-suite for the newly introduced feature of
ARM Dynamic Direct Interrupts.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: introduce support for dynamic direct interrupts

With this commit we add support for Dynamic Direct interrupts
for the ARM Cortex-M architecture. For that we introduce a new,
user-enabled, Kconfig symbol, DYNAMIC_DIRECT_INTERRUPTS.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: cortex-m: implement support for dynamic direct interrupts

This commits implements the support for dynamic direct
interrupts for the ARM Cortex-M architecture, and exposes
the support to the user as an ARM-only API.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

carlescufi commented on tests/arch/arm/arm_irq_advanced_features/src/arm_dynamic_direct_interrupts.c:
```diff
@@ -0,0 +1,42 @@
+{
+  ARG_UNUSED(args);
+
+	printk("test\n");
```
Why not set a variable and then check that it is set in the test main function?

ioannisg replied:
Work in progress - stand by!

joerchan commented on tests/arch/arm/arm_irq_advanced_features/src/arm_dynamic_direct_interrupts.c:
```diff
@@ -0,0 +1,42 @@
+{
+  /* Place the dynamic interrupt dispatcher in the ROM ISR table. */
+	IRQ_DIRECT_CONNECT(TIMER0_IRQn, 0,
+		z_arm_irq_direct_dynamic_dispatch_no_reschedule, 0);
```
Perhaps we could create a simpler way to declare this ISR?

Now the application has to reference the ISR direct dynamic wrapper. Instead we could just have something that declared TIMER0_IRQn has an interrupt that needs to be connected dynamically.
That is what I wanted to solve with this:
https://github.com/zephyrproject-rtos/zephyr/pull/20387/files#diff-647d5b552bb42878dc0504d70ee3ac50R56

Maybe the tests have to be written this way, but the application should have something that doesn't reference reference zephyr internal functions.

ioannisg replied:
It does not reference internal function; it references an ARM-only API.
I am OK with removing the z_ prefix, to indicate _better_ that this is not internal function. But, somehow, this is an architecture-specific API, so maybe call it simply: `arm_irq_direct_dynamix_dispatch_no_reschedule`?

@carlescufi what is the convention for ARCH-specific APIs exposed, though, publicly?

carlescufi replied:
>@carlescufi what is the convention for ARCH-specific APIs exposed, though, publicly?

Not an expert here, @andrewboie is the person to ask, but from what I can tell functions in `include/arch/` seem to be prefixed with `arch_`

ioannisg replied:
Alright

carlescufi commented on arch/arm/core/irq_manage.c:
```diff
@@ -262,4 +262,33 @@ int arch_irq_connect_dynamic(unsigned int irq, unsigned int priority,
+	if (irq < IRQ_TABLE_SIZE) {
+		struct _isr_table_entry *isr_entry = &_sw_isr_table[irq];
+
+		isr_entry->isr(isr_entry->arg);
```
Why not do:
`return isr_entry->isr(isr_entry->arg);`
and then you don't need the `_reschedule` and `_no_reschedule` variants at all

ioannisg replied:
```
struct _isr_table_entry {
	void *arg;
	void (*isr)(void *);
};
```

Looks to me function returns void

ioannisg committed:
doc: list dynamic direct interrupts as ARM supported API

Document that the Dynamic Direct interrupts feature is
implemented and supported as an ARM-only API.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

carlescufi commented on include/arch/arm/irq.h:
```diff
@@ -136,6 +137,80 @@ static inline void arch_isr_direct_footer(int maybe_swap)
+ *
+ * Note: the function is an ARM Cortex-M only API.
+ */
+#define ARM_IRQ_DIRECT_DYNAMIC_CONNECT_RESCHEDULE(irq_p, priority_p, flags_p) \
```
Can we have a single macro with an additional flag:

```
#define ARM_IRQ_DIRECT_DYNAMIC_CONNECT(irq_p, priority_p, flags_p, reschedule)
IRQ_DIRECT_CONNECT(irq_p, priority_p, \
		z_arm_irq_direct_dynamic_dispatch_reschedule, flags_p | (IRQ_FLAG_RESCHEDULE & reschedule))
```
or something similar?

ioannisg replied:
We can play with the pre-processor and the dispatcher name, stgh like:

```
#define ARM_IRQ_DIRECT_DYNAMIC_CONNECT(irq_p, priority_p, flags_p, reschedule_request)
IRQ_DIRECT_CONNECT(irq_p, priority_p, \
		z_arm_irq_direct_dynamic_dispatch_##reschedule_request, flags_p | (IRQ_FLAG_RESCHEDULE & reschedule))
```

But then we should ensure the macro to be used with certain arguments.
Not sure if it makes it look nicer at the end, @carlescufi

carlescufi replied:
Isn't there a way to check the flags at runtime from a single dispatch function?
If there isn't I still think your proposal above is better than documenting 2 separate macros.

ioannisg replied:
Eventually, we need to define at compile-time two dispatch functions returning true or false.
So there's no way around that.

OK, I can play with the pre-processor and document one single macro. Will try this and update the PR tomorrow. But the limitation is that the user then _must_ use the macro with certain _string_ arguments, otherwise there will be compilation errors. I don't know if we want that.

ioannisg replied:
done @carlescufi

carlescufi commented on tests/arch/arm/arm_irq_advanced_features/src/arm_dynamic_direct_interrupts.c:
```diff
@@ -0,0 +1,86 @@
+	ARM_IRQ_DIRECT_DYNAMIC_CONNECT(DIRECT_ISR_OFFSET, 0, 0, no_reschedule);
+
+	/* Attach the ISR handler at run time. */
+	arch_irq_connect_dynamic(DIRECT_ISR_OFFSET, 0 /* highest priority */,
```
should this instead use `irq_connect_dynamic`?

ioannisg replied:
done

carlescufi commented on include/arch/arm/irq.h:
```diff
@@ -136,6 +137,49 @@ static inline void arch_isr_direct_footer(int maybe_swap)
+extern void z_arm_irq_direct_dynamic_dispatch_no_reschedule(void);
+
+/**
+ * @brief Macro to register an ISR Dispatcher (with/without re-scheduling
```
One thing that you are not documenting here is whether the user should use `IRQ_DIRECT_DECLARE` or just use a `void isr(void)` function instead.

carlescufi replied:
As additional info, direct interrupts are declared as `static inline int name##_body(void);`

carlescufi reviewed (changes requested):
LGTM now, thank you.

carlescufi commented on arch/arm/core/irq_manage.c:
```diff
@@ -262,4 +262,33 @@ int arch_irq_connect_dynamic(unsigned int irq, unsigned int priority,
+	}
+}
+
+ARCH_ISR_DIRECT_DECLARE(z_arm_irq_direct_dynamic_dispatch_reschedule)
```
Why not use `ISR_DIRECT_DECLARE` here instead?

ioannisg replied:
I could, but here it does not matter so much; this is an ARCH ARM test, so in theory we are testing things below the public (kernel) API :)
But I can change it.

ioannisg replied:
done now.

carlescufi commented on include/arch/arm/irq.h:
```diff
@@ -136,6 +137,49 @@ static inline void arch_isr_direct_footer(int maybe_swap)
+extern void z_arm_irq_direct_dynamic_dispatch_no_reschedule(void);
+
+/**
+ * @brief Macro to register an ISR Dispatcher (with/without re-scheduling
```
Here is my take on this:

```

/**
 * @brief Macro to register an ISR Dispatcher (with or without re-scheduling
 * request) for dynamic direct interrupts.
 *
 * This macro registers the ISR dispatcher function for dynamic direct
 * interrupts, allowing the use of dynamic direct ISRs in the kernel.
 * The dispatcher function is invoked when the direct
 * interrupt occurs and then triggers the (soft) Interrupt Service Routine
 * (ISR) that is registered dynamically (i.e. at run-time) into the IRQ table
 * stored in SRAM. The ISR must be connected with irq_connect_dynamic() and
 * enabled via irq_enable() before the dynamic direct interrupt can be
 * serviced. This ISR dispatcher may be configured by the user to trigger
 * thread re-secheduling upon return.
 *
 * These ISRs are designed for performance-critical interrupt handling and do
 * not go through all of the common interrupt handling code. 
 *
 * Dynamic 'direct' interrupts are regular Zephyr interrupts when it comes to their
 * declaration: their signature must match void isr(void* parameter), since, unlike
 * regular direct interrupts, they are not placed directly into the hardware vector
 * table but instead use the software ISR table.
 * The major difference with regular Zephyr interrupts are the following:
 * - Just like direct interrupts, the call into the OS 
 *   to exit power management idle state is optional. Normal interrupts always do this 
 *   before the ISR is run, but when it runs is controlled by the placement of 
 *   a ISR_DIRECT_PM() macro, or omitted entirely.
 * - Similar to direct interrupts, scheduling decisions are optional but must be 
 *   made at build time. They are controlled by @param resch to this macro. 
 *
 *
 * @param irq_p IRQ line number.
 * @param priority_p Interrupt priority.
 * @param flags_p Architecture-specific IRQ configuration flags.
 * @param resch Set flag to 'reschedule' to request thread
 *              re-scheduling upon ISR function. Set flag
 *              'no_reschedule' to skip thread re-scheduling
 *
 * Note: the function is an ARM Cortex-M only API.
 *
 * @return Interrupt vector assigned to this interrupt.
 */
```

carlescufi commented on include/arch/arm/irq.h:
```diff
@@ -136,6 +137,58 @@ static inline void arch_isr_direct_footer(int maybe_swap)
+extern void z_arm_irq_direct_dynamic_dispatch_reschedule(void);
+extern void z_arm_irq_direct_dynamic_dispatch_no_reschedule(void);
+
+/**
```
@ioannisg 
Some fixes:

```
/**
 * @brief Macro to register an ISR Dispatcher (with or without re-scheduling
 * request) for dynamic direct interrupts.
 *
 * This macro registers the ISR dispatcher function for dynamic direct
 * interrupts for a particular IRQ line, allowing the use of dynamic 
 * direct ISRs in the kernel for that interrupt source.
 * The dispatcher function is invoked when the hardware
 * interrupt occurs and then triggers the (software) Interrupt Service Routine
 * (ISR) that is registered dynamically (i.e. at run-time) into the software ISR table
 * stored in SRAM. The ISR must be connected with irq_connect_dynamic() and
 * enabled via irq_enable() before the dynamic direct interrupt can be
 * serviced. This ISR dispatcher must be configured by the user to trigger
 * thread re-secheduling upon return, using the @param resch parameter.
 *
 * These ISRs are designed for performance-critical interrupt handling and do
 * not go through all of the common interrupt handling code.
 *
 * With respect to their declaration, dynamic 'direct' interrupts are regular
 * Zephyr interrupts; their signature must match void isr(void* parameter), as,
 * unlike regular direct interrupts, they are not placed directly into the
 * ROM hardware vector table but instead they are installed in the software ISR table.
 *
 * The major differences with regular Zephyr interrupts are the following:
 * - Similar to direct interrupts, the call into the OS to exit power
 *   management idle state is optional. Normal interrupts always do this
 *   before the ISR is run, but with dynamic direct ones when and if it runs 
 *   is controlled by the placement of
 *   a ISR_DIRECT_PM() macro, or omitted entirely.
 * - Similar to direct interrupts, scheduling decisions are optional. Unlike
 *   direct interrupts, the decision must be made at build time. 
 *   They are controlled by @param resch to this macro.
 *
 * @param irq_p IRQ line number.
 * @param priority_p Interrupt priority.
 * @param flags_p Architecture-specific IRQ configuration flags.
 * @param resch Set flag to 'reschedule' to request thread
 *              re-scheduling upon ISR function. Set flag
 *              'no_reschedule' to skip thread re-scheduling
 *
 * Note: the function is an ARM Cortex-M only API.
 *
 * @return Interrupt vector assigned to this interrupt.
 */
```

carlescufi commented:
@ioannisg please rebase

ioannisg commented:
> @ioannisg please rebase

@carlescufi rebased :)

carlescufi commented:
@ioannisg there seems to be an issue with shippable. Perhaps it just needs a rebase.

dbkinder reviewed (approved):
+1 for doc changes

andrewboie commented on arch/arm/core/aarch32/cortex_m/Kconfig:
```diff
@@ -243,6 +243,14 @@ config ZERO_LATENCY_IRQS
 
+config DYNAMIC_DIRECT_INTERRUPTS
+	bool "Enable support for dynamic direct interrupts"
+	select DYNAMIC_INTERRUPTS
```
@ulfalizer would be the authority but shouldn't we do "depends on DYNAMIC_INTERRUPTS" or "imply DYNAMIC_INTERRUPTS" instead?

"select" is kind of a bludgeon since it ignored dependencies

ulfalizer replied:
If there's a chance that `DYNAMIC_INTERRUPTS` will have dependencies, I'd go with `depends on` at least. Currently it doesn't have any dependencies, so that makes it slightly less bad.

For `imply`, if you do `imply DYNAMIC_INTERRUPTS`, it will still be possible to disable `DYNAMIC_INTERRUPTS` when `DYNAMIC_DIRECT_INTERRUPTS` is enabled. That would be bad if it's required in that case.

andrewboie replied:
I suggest we change this to "depends on" then.

ioannisg replied:
If we change to depends-on, then one would need to set both Kconfig symbols, to have DYNAMIC_DIRECT_INTERRUPTS enabled, correct, @ulfalizer ?

If so, I am not sure I like it, but I am open to ideas :)

tyrantlizard77 reviewed (commented):
No complaints other than a suspicion that the Kconfig isn't quite ideal

ulfalizer commented on arch/arm/core/aarch32/cortex_m/Kconfig:
```diff
@@ -243,6 +243,14 @@ config ZERO_LATENCY_IRQS
+	help
+	  Enable installation of interrupt service routines for Direct
+	  interrupts at runtime.
+	  Note: this enables support for dynamic interrupts in the kernel.
```
Could flesh out this help text a bit maybe. Could explain what direct interrupts are (and maybe lowercase 'direct'), give some links, etc. In the Linux kernel, `checkpatch.pl` warns if the help text is less than four lines, but maybe our version is old.

ioannisg replied:
I see. So, I've changed 'Direct' to lowercase, thanks.

Well, the provided API (see in this PR) has all the documentation for direct dynamic interrupts. So I've just copied here a minor phrase briefly explaining direct interrupts.

In general the help-text of Kconfig symbols does not need to explain in detail the feature/functionality; we've got documentation for that. :)

ioannisg reviewed (commented):
@ulfalizer comments addressed, pls, take another look

ioannisg commented:
@ulfalizer @andrewboie could you re-visit your reviews, here, so we could resolve this one?

andrewboie commented:
Please change the Kconfig to have "depends on" and not "select". Users will need to enable both.

ioannisg commented:
> Please change the Kconfig to have "depends on" and not "select". Users will need to enable both.

That's alright then, @andrewboie , I' ve done that now, so using this feature requires both Kconfigs be set, thanks for looking at it.
